File: r-owino/Week1-Linked_Lists/reverse_list.py
#!/usr/bin/env python3
"""
206. Reverse Linked List

Example:
Input: head = [1,2,3,4,5]
Output: [5,4,3,2,1]
given: 1 -> 2 -> 3 -> 4 -> 5
reverse: 5 -> 4 -> 3 -> 2 -> 1

Input: head = [1,2]
Output: [2,1]
given: 1 -> 2
reverse: 2 -> 1

Input: head = []
Output: []

"""

from typing import Optional, List
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        """reverses a linked list"""
        # check if the list is empty or only has one element
        if head is None or head.next is None:
            return head
        
        # reverse the list
        temp = head
        prev = None

        while temp is not None:
            next_node = temp.next
            temp.next = prev
            prev = temp
            temp = next_node
        
        return prev

def list_to_linked_list(arr: List[int]) -> Optional[ListNode]:
    """convert an input list to a linked list"""
    if not arr:
        return None
    head = ListNode(arr[0])
    curr = head
    for val in arr[1:]:
        curr.next = ListNode(val)
        curr = curr.next
    return head

def linked_list_to_list(head: Optional[ListNode]) -> List[int]:
    result = []
    while head:
        result.append(head.val)
        head = head.next
    return result

sol = Solution()
linked_list = list_to_linked_list([1,2,3,4,5])
print(linked_list_to_list(sol.reverseList(linked_list)))

File: r-owino/Week1-Linked_Lists/reverse_list.rs
// definition for a singly-linked list
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
   pub val: i32,
   pub next: Option<Box<ListNode>>,
}

impl ListNode {
   #[inline]
   fn new(val: i32) -> Self {
       ListNode {
           next: None,
           val
       }
   }
}

struct Solution;

impl Solution {
    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut prev: Option<Box<ListNode>> = None;
        let mut curr = head;

        while let Some(mut node) = curr {
            curr = node.next.take();
            node.next = prev;
            prev = Some(node);
        }
        prev
    }
}

// convert Vec<i32> to Option<Box<ListNode>>
fn vec_to_linked_list(values: Vec<i32>) -> Option<Box<ListNode>> {
    let mut curr = None;
    for &val in values.iter().rev() {
        let mut new_node = Box::new(ListNode::new(val));
        new_node.next = curr;
        curr = Some(new_node);
    }
    curr
}

// convert Option<Box<ListNode>> to Vec<i32>
fn linked_list_to_vec(mut head: Option<Box<ListNode>>) -> Vec<i32> {
    let mut values = vec![];
    while let Some(node) = head {
        values.push(node.val);
        head = node.next;
    }
    values
}

fn main() {
    let head = vec_to_linked_list(vec![4]);
    let reversed_vec = linked_list_to_vec(Solution::reverse_list(head));
    println!("{:?}", reversed_vec);
}
